src/storage/buffer/buffer_manager.cpp

```diff
@@ -45,106 +47,101 @@ BufferObj *BufferManager::Allocate(UniquePtr<FileWorker> file_worker) {
+void BufferManager::RemoveClean() {
+    Vector<BufferObj *> clean_list;
+    {
+        std::unique_lock lock(clean_locker_);
+        clean_list.swap(clean_list_);
     }
 
     {
-        std::unique_lock<std::mutex> lock(w_locker_);
-        for (SizeT i = 0; i < obj_path_delete_.size(); i++) {
-            const auto &file_path = obj_path_delete_[i];
-            LOG_TRACE(fmt::format("Erasing buffer map: {}", file_path));
+        std::unique_lock lock(gc_locker_);
+        for (auto *buffer_obj : clean_list) {
+            if (auto iter = gc_map_.find(buffer_obj); iter != gc_map_.end()) {
+                gc_list_.erase(iter->second);
+                gc_map_.erase(iter);
+            }
+        }
+    }
+    {
+        std::unique_lock lock(w_locker_);
+        for (auto *buffer_obj : clean_list) {
+            auto file_path = buffer_obj->GetFilename();
             buffer_map_.erase(file_path);
-            LOG_TRACE(fmt::format("Erased buffer map: {}", file_path));
         }
     }
-
-    // clear
-    file_path_delete_.clear();
-    obj_path_delete_.clear();
 }
 
-void BufferManager::RequestSpace(SizeT need_size, BufferObj *buffer_obj) {
-    while (current_memory_size_ + need_size > memory_limit_) {
-        BufferObj *buffer_obj1 = nullptr;
-        /// FIXME: Actually it is not a good idea to use the moodycamel::ConcurrentQueue.
-        /// When dealing with heavy concurrent read requests, and each request holds some buffer_objs that may cause
-        /// memory usage to exceed the memory_limit, the current handling will lead to some problems. We should consider
-        /// triggering the garbage collection (gc) to free up unnecessary buffer_objs in gc_queue. But currently, we
-        /// only attempt this once, and the service may crash if the head item of the queue shouldn't be freed.
-        /// TODO:
-        /// 1. Replace moodycamel::ConcurrentQueue with a queue using a replacement strategy.
-        /// 2. When encountering out-of-memory situations, first attempt to trigger the gc to try to free up some buffer_objs
-        ///    (not only once now) in gc_queue.
-        /// 3. If the memory usage still exceeds the memory_limit, then consider crashing the service or outputting a warning log.
-        if (gc_queue_.TryDequeue(buffer_obj1)) {
-            if (buffer_obj == buffer_obj1) {
-                UnrecoverableError("buffer object duplicated in gc_queue.");
-            }
-            auto size = buffer_obj1->GetBufferSize();
-            LOG_TRACE(fmt::format("Free buffer object: {}", buffer_obj1->GetFilename()));
-            if (buffer_obj1->Free()) {
-                current_memory_size_ -= size;
-            }
-            LOG_TRACE("Freed buffer object");
-        } else {
-            UnrecoverableError("Out of memory.");
+void BufferManager::RequestSpace(SizeT need_size) {
+    std::unique_lock lock(gc_locker_);
+    while (current_memory_size_ + need_size > memory_limit_ && !gc_list_.empty()) {
```

Comment:
current_memory_size_  is still atomic var, the locker can only protect the while loop body.